go/mysql/mysql56_gtid_set.go

```diff
@@ -334,6 +334,83 @@ func (set Mysql56GTIDSet) AddGTID(gtid GTID) GTIDSet {
 	return newSet
 }
 
+// Union implements GTIDSet.Union().
+func (set Mysql56GTIDSet) Union(other GTIDSet) GTIDSet {
+	if set == nil || other == nil {
+		return set
+	}
+	mydbOther, ok := other.(Mysql56GTIDSet)
+	if !ok {
+		return set
+	}
+
+	// Make a copy and add the new GTID in the proper place.
+	// This function is not supposed to modify the original set.
+	newSet := make(Mysql56GTIDSet)
+
+	for OtherSID, otherIntervals := range mydbOther {
+		intervals, ok := set[OtherSID]
+
+		if !ok {
+			// No matching server id, so we must add it from other set.
+			newSet[OtherSID] = otherIntervals
+			continue
+		}
+
+		// Found server id match between sets, so now we need to add each interval.
+		s1 := intervals[:]
+		s2 := otherIntervals[:]
+		var nextInterval interval
+		var newIntervals []interval
+
+		// While our stacks have intervals to process, do work.
+		for len(s1) != 0 || len(s2) != 0 {
+			// Find which intervals list has earliest start.
+			if len(s2) == 0 || (len(s1) != 0 && s1[0].start <= s2[0].start) {
+				nextInterval = s1[0]
+				// Progress pointer since this stack has earliest.
+				s1 = s1[1:]
+			} else {
+				nextInterval = s2[0]
+				// Progress pointer since this stack has earliest.
+				s2 = s2[1:]
+			}
+
+			if len(newIntervals) == 0 {
+				newIntervals = append(newIntervals, nextInterval)
+			}
+
+			activeInterval := &newIntervals[len(newIntervals)-1]
+
+			if nextInterval.end < activeInterval.end {
+				// We hit an interval whose start was after the previous intervals start, but whose
+				// end is prior to the active intervals end. Skip to next interval.
+				continue
+			}
+
+			if nextInterval.start > activeInterval.end {
```

Comment:
Both sides of the interval are inclusive, so we would extend if it's one past the end.

```suggestion
			if nextInterval.start > activeInterval.end + 1 {
```

Reply:
Ahh, good to know. Will update.

Reply:
Added a unit test to catch this as well, and then fixed. Done.